File: api/src/graphql/user/mutations.ts
import { singleQuote } from "./../common/helper";

// accept friend request messages
export const friendAddedSuccess = singleQuote("Zaakceptowano zaproszenie do znajomych!");
const friendDeletedSuccess = singleQuote("Usunięto znajomego!");

export const friendshipRequestSentSuccess = singleQuote("Wysłano zaproszenie do znajomych!");
export const friendshipRequestDeniedSuccess = singleQuote("Odrzucono prośbę o dołączenie do znajomych!");

const deleteAccountSuccess = singleQuote("Konto zostało usunięte!");
const updatedLocationSuccess = singleQuote("Zaktualizowano lokalizacje!");

export const AccountMutations = `
    acceptFriendRequest(input: FriendInput!): Message
    @cypher(
    statement: """
        MATCH(a: Account { id: $meId })
        MATCH(b: Account { id: $input.id })
        MATCH (a)<-[f:REQUESTED_FRIENDS]-(b)
        DELETE f
        MERGE(a)-[:FRIENDS]->(b)
        MERGE(b)-[:FRIENDS]->(a)
        RETURN {status: 'OK', message: ${friendAddedSuccess}}

    """
    )

    removeFriend(input: FriendInput!): Message
    @cypher(
    statement: """
        MATCH(a: Account { id: $meId })
        MATCH(b: Account { id: $input.id })
        OPTIONAL MATCH (a)-[fa:FRIENDS]->(b)
        OPTIONAL MATCH (b)-[fb:FRIENDS]->(a)
        DELETE fa, fb
        RETURN {status: 'OK', message: ${friendDeletedSuccess}}
    """
    )

    sendFriendRequest(input: FriendInput!): Message
    @cypher(
    statement: """
        MATCH(a: Account { id: $meId })
        MATCH(b: Account { id: $input.id })
        CALL apoc.do.case([
            EXISTS((b)-[:REQUESTED_FRIENDS]->(a))=true,
                \\" WITH $a AS a, $b AS b
                OPTIONAL MATCH(a)-[fa:REQUESTED_FRIENDS]->(b)
                OPTIONAL MATCH(b)-[fb:REQUESTED_FRIENDS]->(a)
                DELETE fa, fb
                MERGE(a)-[:FRIENDS]->(b)
                MERGE(b)-[:FRIENDS]->(a)
                RETURN {status: 'OK', message: ${friendAddedSuccess}} AS result \\",
            EXISTS((b)-[:REQUESTED_FRIENDS]->(a))=false,
                \\"  WITH $a AS a, $b AS b
                MERGE(a)-[:REQUESTED_FRIENDS]->(b)
                RETURN {status: 'OK', message: ${friendshipRequestSentSuccess}} AS result \\"
        ],
        '',
        {a:a, b:b}
        ) YIELD value
        RETURN value.result
    """
    )

    denyFriendRequest(input: FriendInput!): Message
    @cypher(
    statement: """
        MATCH(a: Account { id: $meId })
        MATCH(b: Account { id: $input.id })
        MATCH (a)<-[f:REQUESTED_FRIENDS]-(b)
        DELETE f
        RETURN {status: 'OK', message: ${friendshipRequestDeniedSuccess}}
    """
    )

    updateAccount(input: UpdateAccountInput!): User
    @cypher(
        statement: """
            MATCH(a: Account { id: $meId })
            SET a += apoc.map.fromLists($keys,$values)
            RETURN a
        """
    )

    register(input: RegisterAccountInput): Account
    @cypher(
    statement: """
        CREATE (u:Account {
            name: $input.name,
            email: $input.email,
            password: $input.password,
            age: $input.age,
            favoriteAlcoholName: $input.favoriteAlcoholName,
            favoriteAlcoholType: $input.favoriteAlcoholType,
            description: $input.description,
            gender: $input.gender,
            genderPreference: $input.genderPreference,
            agePreferenceFrom: $input.agePreferenceFrom,
            agePreferenceTo: $input.agePreferenceTo,
            alcoholPreference: $input.alcoholPreference
        })
        CREATE (g: Group)
        SET u.id = apoc.create.uuid()
        SET g.id = apoc.create.uuid()
        MERGE(u)-[:OWNER]->(g)
        RETURN u
    """
    )

    login(input: LoginInput!): LoginPayload

    deleteAccount: Message
    @cypher(
        statement: """
            MATCH (a:Account {id: $meId})
            MATCH (a)-[:OWNER]-(g:Group)
            MATCH (a)-[relations]-(any)
            DELETE relations, a, g
            RETURN {status: 'OK', message: ${deleteAccountSuccess}}
        """
    )

    updateLocation(input: UpdateLocationInput): Message
    @cypher(
        statement: """
            MATCH (a:Account {id: $meId})
            SET a.latestLocation = $location
            RETURN {status: 'OK', message: ${updatedLocationSuccess}}
        """
    )
`;

File: data_generator/mutations.py
from gql import gql

register = gql("""
mutation Register(
    $name: String!,
    $email: String!,
    $password: String!,
    $latitude: Float!,
    $longitude: Float!,
    $age: Int!,
    $favoriteAlcoholName: String!,
    $favoriteAlcoholType: AlcoholType!,
    $description: String!,
    $gender: Gender!,
    $genderPreference: Gender,
    $alcoholPreference: AlcoholType!,
    $agePreferenceFrom: Int!,
    $agePreferenceTo: Int!,
  ) {
  register(input: {
    name: $name,
    email: $email,
    password: $password,
    latitude: $latitude,
    longitude: $longitude,
    age: $age,
    favoriteAlcoholName: $favoriteAlcoholName,
    favoriteAlcoholType: $favoriteAlcoholType,
    description: $description,
    gender: $gender,
    genderPreference: $genderPreference,
    alcoholPreference: $alcoholPreference,
    agePreferenceFrom: $agePreferenceFrom,
    agePreferenceTo: $agePreferenceTo,
  }) {
    id
  }
}
""")

login = gql("""
mutation Login($email: String!, $password: String!) {
  login(input: {
    email: $email,
    password: $password
  }) {
    token
  }
}
""")

acceptFriendRequest = gql("""
mutation AcceptFriendRequest($input: FriendInput!){
    acceptFriendRequest(input: $input) {
      message
      status
    }
  }
""")

denyFriendRequest = gql("""
mutation DenyFriendRequest($input: FriendInput!){
    denyFriendRequest(input: $input) {
      message
      status
    }
}
""")

removeFriend = gql("""
mutation RemoveFriend($input: FriendInput!){
    removeFriend(input: $input) {
      message
      status
    }
  }
""")

addFriend = gql("""
mutation SendFriendRequest($input: FriendInput!){
    sendFriendRequest(input: $input) {
      message
      status
    }
  }
""")


updateLocation = gql("""
mutation UpdateLocation($latitude: Float, $longitude: Float){
  updateLocation(input: {
    latitude: $latitude,
    longitude: $longitude}){
      message
      status
    }
}
""")

acceptGroupRequest = gql("""
mutation AcceptGroupInvitation($input: AcceptGroupInvitationInput!){
   acceptGroupInvitation(input: $input) {
      message
      status
    }
  }
""")

rejectGroupRequest = gql("""
mutation RejectGroupInvitation($input: RejectGroupInvitationInput!){
    rejectGroupInvitation(input: $input) {
      message
      status
    }
}
""")

createGroup = gql("""
mutation CreateGroup($usersId: [ID!]!){
    createGroup(input: {
    usersId: $usersId
    }) {
      message
      status
    }
}
""")

addToGroup = gql("""
mutation AddFriendToGroup($groupId: ID!, $friendId: ID!){
    addFriendToGroup(input: {
    friendId: $friendId,
    groupId: $groupId
    }) {
      message
      status
    }
}
""")

voteFor = gql("""
mutation AcceptGroupPendingUser($groupId: ID!, $userId: ID!){
    acceptGroupPendingUser(input: {
    userId: $userId,
    groupId: $groupId
    }) {
      message
      status
    }
}
""")

voteAgainst = gql("""
mutation RejectGroupPendingUser($groupId: ID!, $userId: ID!){
    rejectGroupPendingUser(input: {
    userId: $userId,
    groupId: $groupId
    }) {
      message
      status
    }
}
""")

sendChatMessage = gql("""
mutation SendChatMessage($groupId: ID!, $message: String!) {
  sendChatMessage(input: {
    groupId: $groupId, message: $message
  }) {
    status
    message
  }
}
""")

swipeToLike = gql("""
mutation SwipeToLike($groupId: ID!){
    swipeToLike(input: {
    groupId: $groupId
    }) {
      message
      status
    }
}
""")

swipeToDislike = gql("""
mutation SwipeToDislike($groupId: ID!){
    swipeToDislike(input: {
    groupId: $groupId
    }) {
      message
      status
    }
}
""")